### PR TITLE
drivers: sdhc: sdhc_spi: release bus on error, and fix CMD12 failue logic

### LINKED ISSUE #72364: drivers: sdhc: Why does sdhc_spi_request not spi_release() on error?
**Describe the bug**
Suspected bug: [sdhc_spi_request()](https://github.com/nrfconnect/sdk-zephyr/blob/b5cf3040298410845311ff7397374540d3451528/drivers/sdhc/sdhc_spi.c#L600) calls spi_release() at its end.  
However, if it returns an error, the function does not release the SPI. 

**Expected behavior**
The current behaviour assumes that on an error, other parts of the SPI would have either not locked SPI, or properly release the SPI on errors.  Instead, would it not be more robust to release the SPI when the sdhc_spi_request exists, regardless of error?

**Context**
I work with Nordic Semiconductor DevZone, and a user asked me about this over there. Private ticket so I can not link it here.

### LINKED ISSUE #72365: drivers: sdhc: Not checking CMD12 return
**Describe the bug**
If sdhc_spi_request() tries to send a cmd and fails, it may do [CMD12](https://github.com/nrfconnect/sdk-zephyr/blob/b5cf3040298410845311ff7397374540d3451528/drivers/sdhc/sdhc_spi.c#L630-L637). However, the return from this second send command is not handled. If that second CMD12 also fails, it will not be caught by the driver.

**Expected behavior**
The CMD12 should have error handling.

**Context**
I work with Nordic Semiconductor DevZone, and a user asked me about this over there. Private ticket so I can not link it here.
The user I have reports that they see the send failing twice.

### PR DESCRIPTION
**Pull request by danieldegrasse**
Properly release SPI bus on transmit error within the SDHC SPI driver. In these cases return code is not checked, as we wish to return the error code from the failed transfer to the SD stack.

Fixes #72364

drivers: sdhc: sdhc_spi: rework CMD12 failure logic

Rework CMD12 failure logic for SDHC SPI driver. Previously, the error
code of CMD12 was not checked, so even if CMD12 failed to send the
initial command would be retried. Change this behavior to retry CMD12
until it succeeds. If CMD12 fails, its error code will be propagated to
the caller. Otherwise, the return code from the command being sent by
the caller will be propagated.

Fixes #72365

### CODE REVIEW

File: drivers/sdhc/sdhc_spi.c [hunk 1]
```diff
@@ -631,13 +633,22 @@ static int sdhc_spi_request(const struct device *dev,
 				/* CMD12 is required after multiple read, or
 				 * to retry failed transfer
 				 */
-				sdhc_spi_send_cmd(dev,
+				stop_ret = sdhc_spi_send_cmd(dev,
 					(struct sdhc_command *)&stop_cmd,
 					false);
+				while ((stop_ret != 0) && (retries < 0)) {
```
Comment (decsny) [inline drivers/sdhc/sdhc_spi.c:639]: why is it retries < 0?

File: drivers/sdhc/sdhc_spi.c [hunk 2]
```diff
@@ -631,13 +633,22 @@ static int sdhc_spi_request(const struct device *dev,
...
+					/* Retry stop command */
+					ret = stop_ret = sdhc_spi_send_cmd(dev,
+						(struct sdhc_command *)&stop_cmd,
+						false);
+					retries--;
```
Comment (decsny) [inline drivers/sdhc/sdhc_spi.c:644]: changing the same variable retries in so many places is making this code confusing to read for me

File: drivers/sdhc/sdhc_spi.c (hunk 1 above)
Comment (danieldegrasse) [reply drivers/sdhc/sdhc_spi.c:639]: Good catch, this should be `retries > 0`.

File: drivers/sdhc/sdhc_spi.c (hunk 2 above)
Comment (danieldegrasse) [reply drivers/sdhc/sdhc_spi.c:644]: Is there a way you'd rather structure this? As far as I know, only two locations in this function decrement the value

Comment (decsny) [reply drivers/sdhc/sdhc_spi.c:644]: i guess it's the changing and the checking, it feels like this code is doing 2 different things with one variable, there is retries for sending the stop command and retries for the request as a whole, should these be two different counts?

Comment (danieldegrasse) [reply drivers/sdhc/sdhc_spi.c:644]: Yeah I see the concern- given how we document the `retries` field, having a separate retry counter for this makes sense- I'll add one.

**Review by decsny — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]

**Pull request merged**